lib/web_ui/lib/src/engine/recording_canvas.dart

```diff
@@ -228,12 +228,20 @@ class RecordingCanvas {
   }
 
   void drawDRRect(ui.RRect outer, ui.RRect inner, ui.Paint paint) {
-    // If inner rect is not contained inside outer, flutter engine skips
-    // painting rectangle.
-    if (!(outer.contains(ui.Offset(inner.left, inner.top)) &&
-        outer.contains(ui.Offset(inner.right, inner.bottom)))) {
+    // Ensure inner is fully contained within outer, by comparing its
+    // defining points (including its border radius)
+    final ui.RRect scaled = inner.scaleRadii(); // Measure the same rrect that is going to be painted
+    if (!(outer.contains(ui.Offset(scaled.left, scaled.top + scaled.tlRadiusY))
+        && outer.contains(ui.Offset(scaled.left + scaled.tlRadiusX, scaled.top)) // Left-Top
+        && outer.contains(ui.Offset(scaled.right - scaled.trRadiusX, scaled.top))
+        && outer.contains(ui.Offset(scaled.right, scaled.top + scaled.trRadiusY)) // Right-Top
+        && outer.contains(ui.Offset(scaled.right, scaled.bottom - scaled.brRadiusY))
+        && outer.contains(ui.Offset(scaled.right - scaled.brRadiusX, scaled.bottom)) // Right-Bottom
+        && outer.contains(ui.Offset(scaled.left + scaled.blRadiusX, scaled.bottom))
+        && outer.contains(ui.Offset(scaled.left, scaled.bottom - scaled.blRadiusY)))) { // Left-Bottom
```

Comment:
In the happy case when the drrect is properly specified, which check will allocate 8 `ui.Offset` objects. Is it possible to optimize that away?

Reply:
/cc @rakudrama Am I being too paranoid here?

Reply:
I can try with a different check, but I'm not sure it'll be geometrically precise, let me give that a quick shot, I can always come back to this one.

Reply:
I don't mean rewrite the check. I mean do exactly the same math without allocation. Or is there something that prevents you from doing that?

Reply:
The API of [contains](https://github.com/flutter/engine/blob/master/lib/web_ui/lib/src/ui/geometry.dart#L1502) is what it is, and once the radius information is used to construct the RRect, everything is stored as doubles, so I don't see any way of computing the offsets of each of the points of the corner without allocating an Offset. Also Offsets seem to be immutable, so even if I try to [move it around](https://github.com/flutter/engine/blob/master/lib/web_ui/lib/src/ui/geometry.dart#L232-L233), it'll return a new instance all the time, right?

Anyway, I wrote a slightly different revision that does the same with a little bit less allocation (but in a slightly less compact code).